### DISCUSSION TITLE
Looking to port u8g2 for PIC Microcontroller.  Anybody been there already?

### DISCUSSION THREAD

**Discussion by ChiefEngrNY** [olikraus/u8g2#1884] [Q&A]
Greetings,

I am looking to port the u8g2 driver to run on a MicroChip 16-bit PIC MCU - something in the PIC24 Family.  The PIC development environment compiles C code for the PIC so starting with C is great.

I'm looking to see if anyone has been down this path before - both porting u8g2 to a different MCU as well as trying to get it running on the PIC24.  If so, I'd love to learn from your experiences.

If the price is right, I would be willing to fund the job of porting the package.

Thanks in advance for any assistance and experience anyone might have.

**Comment by olikraus** [comment]
Topic/discussion had been here: https://github.com/olikraus/u8g2/issues/70

So, one precondition is this: You need the professional version of the c-compiler.

If this is the case, you can just go ahead with the instructions from here: https://github.com/olikraus/u8g2/wiki/Porting-to-new-MCU-platform

**Reply by ChiefEngrNY** [reply]
Thank you for the quick reply and apologies for not having searched through the issues;  I did search the discussion topics without a hit.

It sounds promising and with any luck I won't have similar problems.  I'm using the XC16 compiler (since my target is a 16-bit processor) and the device I am planning to use has 128KB of program memory and 16KB of RAM. 

I suppose an interesting test would be to just try to compile the code unchanged to see what kind of errors it throws.

Thanks again!

**Reply by olikraus** [reply]
So you have the paid version of the XC16? It especially requies link time optimization enabled to use the u8g2 c code, otherwise 128KB a flash and 16KB RAM will by far not sufficient ;-)

**Reply by ChiefEngrNY** [reply]
Yikes!  I don't have an active pro license at the moment, but I can renew for a month to do the development work.  I find it more economical to pay for the features only when I need them.

As long as I have you on the line...  Is there a big, fat standalone set of .pdf files of the documentation?  I don't always have network when I am working at some remote locations.

**Reply by olikraus** [reply]
There are only wiki documents.

**Reply by devau** [reply]
Hello
I would like to port usg2 to a PIC16F1825 and then compile it with CCS compiler. @olikraus: The documentation on how to do this is vey complex! I agree with ChiefEngrNY saying "If the price is right, I would be willing to fund the job of porting the package."
I think there are lot of users, not only dealing with Arduino stuff as Microchip's PIC controller are widly used with OLED displays. What do you think about it?

**Reply by olikraus** [reply]
As mentioned, you would require the professional version of the compiler (well I do not know the current situation, but at least it was like this many years ago).

> The documentation on how to do this is vey complex! 

Well, could be, but actually, if you have a I2C display and you know how to clear and set a port, then the HAL code is very simple.

See the HAL code for ARM here: https://github.com/olikraus/u8g2/blob/master/sys/arm/stm32l031x6/u8x8_test/u8x8cb.c

Ok, still the code looks complicated, because the code does all the register updates, but it is really just a little bit of delay and GPIO clear/set functions.

The setup code itself is also simple:
https://github.com/olikraus/u8g2/blob/b1a4917949d80ac0ed59d36721baaed4bbe2d2e4/sys/arm/stm32l031x6/u8x8_test/main.c#L358-L363

Also the magic is done with this call:
```
u8x8_Setup(&u8x8, u8x8_d_ssd1306_128x64_noname, u8x8_cad_ssd13xx_i2c, u8x8_byte_sw_i2c, u8x8_gpio_and_delay_stm32l0);
```
However the only function, which needs to be provided by you is the last callback function: "u8x8_gpio_and_delay_stm32l0" (the one which is linked in the first link above)
All other functions like `u8x8_d_ssd1306_128x64_noname`, `u8x8_cad_ssd13xx_i2c` and `u8x8_byte_sw_i2c` are part of u8g2 / u8x8.

**Reply by devau** [reply]
For sending data over I2C the CCS compiler has the command "**i2c_transfer_out([stream], address, wData, wCount)**".  _stream_ is defined internal, _address_ is the I2C-address of the SSD1306 128x32 display, _wData_ is pointer to data and _wCount_ is number of bytes. Does this mean I have to place this somehow into code "case U8X8_MSG_GPIO_I2C_DATA:" ?  In fact I do not have to set/clear ports directly as this is done all by the library of the CCS-compiler, which addresses itself the internal hardware controlled I2C-module MSSP (see: [https://www.sprut.de/electronic/pic/grund/i2c.htm] part "MSSP-Modul im Master-Mode"). Thanks for your help and greetings from Switzerland :-)

**Reply by olikraus** [reply]
Obviously, you can and should use your internal i2c_transfer_out procedure. In order to do so, you additionally need to reimplement and provide `u8x8_byte_sw_i2c`: Details are here: https://github.com/olikraus/u8g2/wiki/Porting-to-new-MCU-platform#hardware-i2c-communication

The mentioned approach above is more like a first simplified step with as little extra effort from your end. Obviously I2C bit banging will be much slower than using `i2c_transfer_out`. 

Greetings from South Germany ;-)
